File: update-java/src/main/java/com/mqd/updatejava/download/UpdateNotifications.java
package com.mqd.updatejava.download;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;

import com.mqd.updatejava.core.UpdateChecker;

/**
 * 更新通知（不依赖 androidx.core）。
 */
public class UpdateNotifications {

    private static final String CHANNEL_UPDATE = "updatejava_update";
    private static final String CHANNEL_DOWNLOAD = "updatejava_download";

    public static final int NOTIFICATION_ID_NEW_VERSION = 1001;
    public static final int NOTIFICATION_ID_DOWNLOAD = 1002;

    private static void ensureChannels(Context context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return;
        NotificationManager nm = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (nm == null) return;
        if (nm.getNotificationChannel(CHANNEL_UPDATE) == null) {
            NotificationChannel ch = new NotificationChannel(
                    CHANNEL_UPDATE, "更新提醒", NotificationManager.IMPORTANCE_LOW);
            nm.createNotificationChannel(ch);
        }
        if (nm.getNotificationChannel(CHANNEL_DOWNLOAD) == null) {
            NotificationChannel ch = new NotificationChannel(
                    CHANNEL_DOWNLOAD, "下载进度", NotificationManager.IMPORTANCE_LOW);
            nm.createNotificationChannel(ch);
        }
    }

    private static PendingIntent contentIntent(Context context) {
        Intent intent = context.getPackageManager().getLaunchIntentForPackage(context.getPackageName());
        if (intent != null) {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        } else {
            intent = new Intent();
        }
        int flags = PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT;
        return PendingIntent.getActivity(context, 0, intent, flags);
    }

    public static void showNewVersion(Context context, String version) {
        ensureChannels(context);
        Notification.Builder builder;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            builder = new Notification.Builder(context, CHANNEL_UPDATE);
        } else {
            builder = new Notification.Builder(context);
        }
        String contentText = version != null ? UpdateChecker.displayVersion(version) + " 可更新，点击查看" : "发现新版本";
        Notification notification = builder
                .setSmallIcon(getNotificationIcon(context))
                .setContentTitle("发现新版本")
                .setContentText(contentText)
                .setContentIntent(contentIntent(context))
                .setAutoCancel(true)
                .build();
        notifySafely(context, NOTIFICATION_ID_NEW_VERSION, notification);
    }

    public static void cancelNewVersion(Context context) {
        NotificationManager nm = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        if (nm != null) nm.cancel(NOTIFICATION_ID_NEW_VERSION);
    }

    public static Notification buildDownloadNotification(Context context, int progress) {
        ensureChannels(context);
        Notification.Builder builder;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            builder = new Notification.Builder(context, CHANNEL_DOWNLOAD);
        } else {
            builder = new Notification.Builder(context);
        }
        int p = Math.max(0, Math.min(100, progress));
        return builder
                .setSmallIcon(getNotificationIcon(context))
                .setContentTitle("正在下载更新")
                .setContentText(p + "%")
                .setProgress(100, p, false)
                .setOngoing(true)
                .setContentIntent(contentIntent(context))
                .build();
    }

    public static void notifyDownloadProgress(Context context, int progress) {
        notifySafely(context, NOTIFICATION_ID_DOWNLOAD, buildDownloadNotification(context, progress));
    }

    public static void showDownloadDone(Context context, String version) {
        ensureChannels(context);
        Notification.Builder builder;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            builder = new Notification.Builder(context, CHANNEL_DOWNLOAD);
        } else {
            builder = new Notification.Builder(context);
        }
        String contentText = "点击安装 " + UpdateChecker.displayVersion(version);
        Notification notification = builder
                .setSmallIcon(getNotificationIcon(context))
                .setContentTitle("下载完成")
                .setContentText(contentText)
                .setContentIntent(contentIntent(context))
                .setAutoCancel(true)
                .setOngoing(false)
                .build();
        notifySafely(context, NOTIFICATION_ID_DOWNLOAD, notification);
    }

    public static void showDownloadFailed(Context context) {
        ensureChannels(context);
        Notification.Builder builder;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            builder = new Notification.Builder(context, CHANNEL_DOWNLOAD);
        } else {
            builder = new Notification.Builder(context);
        }
        Notification notification = builder
                .setSmallIcon(getNotificationIcon(context))
                .setContentTitle("下载失败")
                .setContentText("点击重试")
                .setContentIntent(contentIntent(context))
                .setAutoCancel(true)
                .setOngoing(false)
                .build();
        notifySafely(context, NOTIFICATION_ID_DOWNLOAD, notification);
    }

    private static void notifySafely(Context context, int id, Notification notification) {
        try {
            NotificationManager nm = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
            if (nm != null) nm.notify(id, notification);
        } catch (Exception ignored) {
        }
    }

    private static int getNotificationIcon(Context context) {
        return context.getResources().getIdentifier(
                "ic_updatelib_notification", "drawable", context.getPackageName());
    }
}

File: update-lib/src/main/java/com/mqd/updatelib/download/UpdateNotifications.kt
package com.mqd.updatelib.download

import android.Manifest
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import com.mqd.updatelib.R
import com.mqd.updatelib.core.UpdateChecker

/**
 * 更新相关通知统一出口：两个 channel + 统一 PendingIntent。
 *
 * - `update` channel：后台「发现新版」提醒。
 * - `download` channel：前台下载进度 / 完成 / 失败（静默 ongoing）。
 *
 * 所有通知点击都拉起应用主界面让 ViewModel 读状态决定弹窗，不依赖 intent 标记。
 * 无 `POST_NOTIFICATIONS` 权限时静默跳过 notify（功能照常，仅无通知）。
 *
 * @author quzhuligpt@gmail.com
 * @since 2026/6/19
 */
object UpdateNotifications {

    private const val CHANNEL_UPDATE = "updatelib_update"
    private const val CHANNEL_DOWNLOAD = "updatelib_download"

    const val NOTIFICATION_ID_NEW_VERSION = 1001
    const val NOTIFICATION_ID_DOWNLOAD = 1002

    private fun ensureChannels(context: Context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
        val nm = context.getSystemService(NotificationManager::class.java) ?: return
        if (nm.getNotificationChannel(CHANNEL_UPDATE) == null) {
            nm.createNotificationChannel(
                NotificationChannel(
                    CHANNEL_UPDATE,
                    context.getString(R.string.updatelib_channel_update),
                    NotificationManager.IMPORTANCE_LOW
                )
            )
        }
        if (nm.getNotificationChannel(CHANNEL_DOWNLOAD) == null) {
            nm.createNotificationChannel(
                NotificationChannel(
                    CHANNEL_DOWNLOAD,
                    context.getString(R.string.updatelib_channel_download),
                    NotificationManager.IMPORTANCE_LOW
                )
            )
        }
    }

    private fun contentIntent(context: Context): PendingIntent {
        val intent = context.packageManager.getLaunchIntentForPackage(context.packageName)?.apply {
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)
        } ?: return PendingIntent.getActivity(
            context, 0, Intent(),
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )
        return PendingIntent.getActivity(
            context,
            0,
            intent,
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )
    }

    /** 后台发现新版时调用。 */
    fun showNewVersion(context: Context, version: String) {
        ensureChannels(context)
        val notification = NotificationCompat.Builder(context, CHANNEL_UPDATE)
            .setSmallIcon(R.drawable.ic_updatelib_notification)
            .setContentTitle(context.getString(R.string.updatelib_notification_new_title))
            .setContentText(context.getString(R.string.updatelib_notification_new_content, UpdateChecker.displayVersion(version)))
            .setContentIntent(contentIntent(context))
            .setAutoCancel(true)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .build()
        notifyIfPermitted(context, NOTIFICATION_ID_NEW_VERSION, notification)
    }

    /** 用户开始下载时去重，取消「发现新版」通知。 */
    fun cancelNewVersion(context: Context) {
        NotificationManagerCompat.from(context).cancel(NOTIFICATION_ID_NEW_VERSION)
    }

    /** 前台下载通知，供 startForeground 与进度刷新复用。 */
    fun buildDownloadNotification(context: Context, progress: Int): Notification {
        ensureChannels(context)
        return NotificationCompat.Builder(context, CHANNEL_DOWNLOAD)
            .setSmallIcon(R.drawable.ic_updatelib_notification)
            .setContentTitle(context.getString(R.string.updatelib_notification_downloading_title))
            .setContentText("$progress%")
            .setProgress(100, progress.coerceIn(0, 100), false)
            .setOngoing(true)
            .setOnlyAlertOnce(true)
            .setContentIntent(contentIntent(context))
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .build()
    }

    fun notifyDownloadProgress(context: Context, progress: Int) {
        notifyIfPermitted(context, NOTIFICATION_ID_DOWNLOAD, buildDownloadNotification(context, progress))
    }

    fun showDownloadDone(context: Context, version: String) {
        ensureChannels(context)
        val notification = NotificationCompat.Builder(context, CHANNEL_DOWNLOAD)
            .setSmallIcon(R.drawable.ic_updatelib_notification)
            .setContentTitle(context.getString(R.string.updatelib_notification_done_title))
            .setContentText(context.getString(R.string.updatelib_notification_done_content, UpdateChecker.displayVersion(version)))
            .setContentIntent(contentIntent(context))
            .setAutoCancel(true)
            .setOngoing(false)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .build()
        notifyIfPermitted(context, NOTIFICATION_ID_DOWNLOAD, notification)
    }

    fun showDownloadFailed(context: Context) {
        ensureChannels(context)
        val notification = NotificationCompat.Builder(context, CHANNEL_DOWNLOAD)
            .setSmallIcon(R.drawable.ic_updatelib_notification)
            .setContentTitle(context.getString(R.string.updatelib_notification_failed_title))
            .setContentText(context.getString(R.string.updatelib_notification_failed_content))
            .setContentIntent(contentIntent(context))
            .setAutoCancel(true)
            .setOngoing(false)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .build()
        notifyIfPermitted(context, NOTIFICATION_ID_DOWNLOAD, notification)
    }

    private fun notifyIfPermitted(context: Context, id: Int, notification: Notification) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
            ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) !=
            PackageManager.PERMISSION_GRANTED
        ) {
            return
        }
        NotificationManagerCompat.from(context).notify(id, notification)
    }

}
